### PR TITLE
Support user-defined key type in JDBC KeyHolder

### PR DESCRIPTION
**Pull request by eXsio**
Enhance JDBC KeyHolder by allowing to retrieve single Generated Key that is not a Number. Generated key can be an UUID or something else. Devs shouldn't have to go manually through complex collections to get it.

### CODE REVIEW

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 1]
```diff
@@ -58,6 +58,21 @@
...
+	/**
+	 * Retrieve the first item from the first map, assuming that there is just
+	 * one item and just one map, and that the item is an instance of provided class.
+	 * This is the typical case: a single, numeric generated key.
+	 * <p>Keys are held in a List of Maps, where each item in the list represents
+	 * the keys for each row. If there are multiple columns, then the Map will have
+	 * multiple entries as well. If this method encounters multiple entries in
+	 * either the map or the list meaning that multiple keys were returned,
+	 * then an InvalidDataAccessApiUsageException is thrown.
+	 * @return the generated key as a number
```
Comment (NathanQingyangXu) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java]: So the javadoc is a little bit inconsistent with the intention of the new method. Still 'numeric key or number' is mentioned in javadoc, but we came up with the new method to overcome that limit, right?

Comment (eXsio) [reply spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java]: uh, of course you're right :) copy pasting error. Fixed.

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 2]
```diff
@@ -58,6 +58,22 @@
 	@Nullable
 	Number getKey() throws InvalidDataAccessApiUsageException;
 
+	/**
+	 * Retrieve the first item from the first map, assuming that there is just
+	 * one item and just one map, and that the item is an instance of provided class.
+	 * This is the typical case: a single, numeric generated key.
```
Comment (NathanQingyangXu) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java]: The above line is still confusing. For such typical case, should we use the previous method returning Number instead?

Comment (eXsio) [reply spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java]: fixed as well.

**Comment by eXsio** [comment]
Can you take a look at this please?

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java [hunk 3]
```diff
@@ -16,15 +16,15 @@
 
 package org.springframework.jdbc.support;
 
+import org.springframework.dao.DataRetrievalFailureException;
+import org.springframework.dao.InvalidDataAccessApiUsageException;
+import org.springframework.lang.Nullable;
+
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java:22]: Please revert the re-ordering of imports.

In addition, please be sure to execute `./gradlew check` before submitting code for a PR in order to ensure that the code does not violate Checkstyle rules.

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java [hunk 4]
```diff
@@ -51,6 +51,7 @@ public GeneratedKeyHolder() {
 
 	/**
 	 * Create a new GeneratedKeyHolder with a given list.
+	 *
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java:54]: Please undo this change.

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java [hunk 5]
```diff
@@ -61,26 +62,34 @@ public GeneratedKeyHolder(List<Map<String, Object>> keyList) {
...
-						"Unable to cast [" + (key != null ? key.getClass().getName() : null) +
-						"] to [" + Number.class.getName() + "]");
+								"Unable to cast [" + key.getClass().getName() +
+								"] to [" + keyClass.getName() + "]");
 			}
-			return (Number) key;
-		}
-		else {
+			return keyClass.cast(key);
+		} else {
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java:92]: In line with Spring's [code style](https://github.com/spring-projects/spring-framework/wiki/Code-Style), please ensure that `else` statements come on the next line.

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java [hunk 6]
```diff
@@ -61,26 +62,34 @@ public GeneratedKeyHolder(List<Map<String, Object>> keyList) {
...
+	@Nullable
+	public <T> T getKeyAs(Class<T> keyClass) throws InvalidDataAccessApiUsageException, DataRetrievalFailureException {
 		if (this.keyList.isEmpty()) {
 			return null;
 		}
 		if (this.keyList.size() > 1 || this.keyList.get(0).size() > 1) {
 			throw new InvalidDataAccessApiUsageException(
 					"The getKey method should only be used when a single key is returned.  " +
-					"The current key entry contains multiple keys: " + this.keyList);
+							"The current key entry contains multiple keys: " + this.keyList);
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java:77]: Please revert this change.

In general, try not to change existing code (including reformatting) if you did not actually change the behavior. Otherwise, such formatting changes add too much noise to the PR and make it harder for people to review the changes.

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java [hunk 7]
```diff
@@ -95,7 +104,7 @@ public Number getKey() throws InvalidDataAccessApiUsageException, DataRetrievalF
 		if (this.keyList.size() > 1) {
 			throw new InvalidDataAccessApiUsageException(
 					"The getKeys method should only be used when keys for a single row are returned.  " +
-					"The current key list contains keys for multiple rows: " + this.keyList);
+							"The current key list contains keys for multiple rows: " + this.keyList);
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/GeneratedKeyHolder.java:107]: Please revert this change (see above for rationale).

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 8]
```diff
@@ -16,21 +16,21 @@
 
 package org.springframework.jdbc.support;
 
-import java.util.List;
-import java.util.Map;
-
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java:21]: Please revert this change (see above for rationale).

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 9]
```diff
@@ -16,21 +16,21 @@
...
 
+import java.util.List;
+import java.util.Map;
+
 /**
  * Interface for retrieving keys, typically used for auto-generated keys
  * as potentially returned by JDBC insert statements.
  *
- * <p>Implementations of this interface can hold any number of keys.
+ * Implementations of this interface can hold any number of keys.
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java:29]: Please revert this change.

The `<p>` is there to begin a new paragraph in the generated Javadoc.

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 10]
```diff
@@ -16,21 +16,21 @@
...
  * In the general case, the keys are returned as a List containing one Map
  * for each row of keys.
  *
- * <p>Most applications only use one key per row and process only one row at a
+ * Most applications only use one key per row and process only one row at a
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java:33]: Please revert this change (see above for rationale).

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 11]
```diff
@@ -47,7 +47,7 @@
 	 * Retrieve the first item from the first map, assuming that there is just
 	 * one item and just one map, and that the item is a number.
 	 * This is the typical case: a single, numeric generated key.
-	 * <p>Keys are held in a List of Maps, where each item in the list represents
+	 * Keys are held in a List of Maps, where each item in the list represents
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java:50]: Please revert this change (see above for rationale).

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 12]
```diff
@@ -58,9 +58,25 @@
...
+	 * @return the generated key as an instance of provided class
+	 * @throws InvalidDataAccessApiUsageException if multiple keys are encountered
+	 */
+	@Nullable
+	<T> T getKeyAs(Class<T> keyClass) throws InvalidDataAccessApiUsageException;
+
 	/**
 	 * Retrieve the first map of keys.
-	 * <p>If there are multiple entries in the list (meaning that multiple rows
+	 * If there are multiple entries in the list (meaning that multiple rows
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java:79]: Please revert this change (see above for rationale).

File: spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java [hunk 13]
```diff
@@ -70,7 +86,7 @@
 
 	/**
 	 * Return a reference to the List that contains the keys.
-	 * <p>Can be used for extracting keys for multiple rows (an unusual case),
+	 * Can be used for extracting keys for multiple rows (an unusual case),
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/support/KeyHolder.java:89]: Please revert this change (see above for rationale).

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for the PR.

I've requested several changes to reduce the scope of the PR and to align with Spring's code style.

It would be great if you could make the changes in the coming days so that this can be included in Spring Framework 5.3 M1 which is scheduled for next week.

**Comment by eXsio** [comment]
@sbrannen thank you for the review. All issues fixed.

**Comment by sbrannen** [comment]
> @sbrannen thank you for the review. 

You're very welcome.

> All issues fixed.

Great. That looks much cleaner now and much easier to review! Speaking of which, I'll make a final pass and merge this in soon.

**Review by sbrannen — APPROVED** [review]

**Comment by sbrannen** [comment]
This has been merged into `master` in b50cf9dad24bd798e28e6fe9293b25910c25bcb4 and slightly revised in cc301011b2edd6c6a02bfe5236482038c4b6473a.

Thanks!

**Comment by sbrannen** [comment]
@eXsio, I'd be happy to add an `@author` tag to `KeyHolder` and `GeneratedKeyHolder` for you, but I don't know your real name.

So if you'd like to be listed as an author and don't mind having your real name published, just let me know...

**Comment by eXsio** [comment]
Hi Sam

That would be great! My name is Slawomir Dymitrow.

Thanks so much!

wt., 16 cze 2020, 14:42 użytkownik Sam Brannen <notifications@github.com>
napisał:

**Comment by sbrannen** [comment]
> Hi Sam That would be great! My name is Slawomir Dymitrow. Thanks so much!

Done in 50391ad3d7a2f16de7466e1cd60b8ac49929fd06. 👍
